File: java/Day5.java
import java.io.*;
import java.util.*;

public class Day5 {
    static String[] mapNames;
    static long[] seeds;
    static Map<String, List<long[]>> mappings; // <mapName, [start, end, diff]>
    
    public static void main(String[] args) throws FileNotFoundException {
        long startTime = System.currentTimeMillis();
        processInput(new Scanner(new FileReader("input/day5.txt")));
        System.out.println("Part I  424490994 : " + getLowestSeed());
        System.out.println("Part II  15290096 : " + getLowestInRange());
        System.out.println("Execution time: " + (System.currentTimeMillis() - startTime) + "ms");
    }
    
    private static long getLowestSeed() {
        long res = Long.MAX_VALUE;
        for (long seed : seeds) {
            for (String mapName : mapNames) {
                seed = getMapTo(mappings.get(mapName), seed);
            }
            res = Math.min(res, seed);
        }
        return res;
    }
    
    private static long getMapTo(List<long[]> ranges, long n) {
        for (long[] range : ranges) {
            if (range[0] <= n && n <= range[1]) {
                return n + range[2];
            }
        }
        return n;
    }
    
    private static long getLowestInRange() {
        List<long[]> mapped = new ArrayList<>();
        for (int i = 0; i < seeds.length - 1; i += 2) {
            mapped.add(new long[]{seeds[i], seeds[i] + seeds[i + 1] - 1});
        }
        for (String mapName : mapNames) {
            mapped = getMapToRange(mappings.get(mapName), mapped);
        }
        long lowest = Long.MAX_VALUE;
        for (long[] range : mapped) {
            lowest = Math.min(lowest, range[0]);
        }
        return lowest;
    }
    
    private static List<long[]> getMapToRange(List<long[]> map, List<long[]> ranges) {
        List<long[]> mapped = new ArrayList<>();
        for (long[] range : ranges) {
            for (long[] mapRange : map) {
                long offBy = mapRange[2];
                if (hasOverlap(range, mapRange)) {
                    mapped.add(new long[]{Math.max(mapRange[0], range[0]) + offBy, Math.min(mapRange[1], range[1]) + offBy});
                }
            }
        }
        return mapped;
    }
    
    private static boolean hasOverlap(long[] range1, long[] range2) {
        return !(range1[1] < range2[0] || range2[1] < range1[0]);
    }
    
    private static void processInput(Scanner in) {
        mapNames = new String[]{"seed-to-soil", "soil-to-fertilizer", "fertilizer-to-water", "water-to-light", "light-to-temperature", "temperature-to-humidity", "humidity-to-location"};
        seeds = Arrays.stream(in.nextLine().split(": ")[1].split(" ")).mapToLong(Long::parseLong).toArray();
        mappings = new HashMap<>();
        for (String key : mapNames) {
            mappings.put(key, new ArrayList<>());
        }
        in.nextLine();
        
        int mapId = 0;
        while (in.hasNextLine()) {
            in.nextLine();
            String key = mapNames[mapId];
            List<long[]> ranges = mappings.get(key);
            
            long lowEnd = Long.MAX_VALUE;
            long highEnd = Long.MIN_VALUE;
            
            while (in.hasNextLine()) {
                String line = in.nextLine();
                if (line.isEmpty()) {
                    mapId++;
                    break;
                }
                String[] entry = line.split(" ");
                long src = Long.parseLong(entry[1]);
                long dst = src + Long.parseLong(entry[2]) - 1; // stupid off by 1
                long offBy = Long.parseLong(entry[0]) - src;
                ranges.add(new long[]{src, dst, offBy});
                
                lowEnd = Math.min(lowEnd, src - 1);
                highEnd = Math.max(highEnd, dst + 1);
            }
            // just trying to cover all ranges
            ranges.add(new long[]{Long.MIN_VALUE, lowEnd, (long) 0});
            ranges.add(new long[]{highEnd, Long.MAX_VALUE, (long) 0});
        }
    }
}

File: rust/src/bin/day5.rs
use std::collections::HashMap;

use lazy_static::lazy_static;

lazy_static!(
    static ref FILE: String = std::fs::read_to_string("../input/day5.txt").unwrap();
);

fn main() {
    let (map_names, seeds, maps) = parse_input(&FILE);
    println!("Part I: Expected 424490994, Result {}", part1(&map_names, &seeds, &maps));
    // println!("Part II: Expected 15290096, Result {}", part2(&map_names, &seeds, &maps));
}

fn part1(map_names: &[&str], seeds: &[i64], maps: &HashMap<&str, Vec<[i64; 3]>>) -> i64 {
    let mut res = i64::MAX;
    for &seed in seeds {
        let mut curr_seed = seed;
        for &name in map_names {
            curr_seed = get_map_to(curr_seed, maps.get(name).unwrap());
        }
        res = res.min(curr_seed);
    }
    res
}

fn get_map_to(seed: i64, ranges: &Vec<[i64; 3]>) -> i64 {
    for &range in ranges {
        if range[0] <= seed && seed <= range[1] {
            return seed + range[2];
        }
    }
    seed
}


fn parse_input(input: &str) -> (Vec<&str>, Vec<i64>, HashMap<&str, Vec<[i64; 3]>>) {
    let map_names = ["seed-to-soil", "soil-to-fertilizer", "fertilizer-to-water", "water-to-light", "light-to-temperature", "temperature-to-humidity", "humidity-to-location"];

    let mut lines = input.lines();
    let seeds = lines.next().unwrap().split(": ").nth(1).unwrap().split_whitespace().map(|s| s.parse::<i64>().unwrap()).collect();
    let mut maps: HashMap<&str, Vec<[i64; 3]>> = HashMap::new();
    for key in map_names {
        maps.insert(key, Vec::new());
    }

    lines.next();
    let mut map_id = 0;
    while let Some(_) = lines.next() {
        // lines.next();
        let key = map_names[map_id];
        let ranges = maps.get_mut(key).unwrap();

        let mut low = i64::MAX;
        let mut high = i64::MIN;

        while let Some(line) = lines.next() {
            if line.is_empty() {
                map_id += 1;
                break;
            }
            let entry: Vec<i64> = line.split_whitespace().map(|s| s.parse::<i64>().unwrap()).collect();
            let src = entry[1];
            let dst = src + entry[2] - 1;
            let off_by = entry[0] - src;
            ranges.push([src, dst, off_by]);

            low = low.min(src - 1);
            high = high.max(dst + 1);
        }
        ranges.push([i64::MIN, low, 0]);
        ranges.push([high, i64::MAX, 0]);
    }

    (map_names.to_vec(), seeds, maps)
}
